app/controllers/base_opinion_controller.rb

```diff
@@ -0,0 +1,34 @@
+class BaseOpinionController < ApplicationController
+  before_action :check_for_opinion_collision, only: :create
+
+  def create
+    @opinion = current_user.send(opinion_type).new(permitted_params)
+
+    if !@opinion.save
+      flash[:notice] = @opinion.errors.full_messages.to_sentence
+    end
+
+    redirect_back(fallback_location: articles_path)
+  end
+
+  def destroy
+    @opinion = current_user.send(opinion_type).find(params[:id])
+    @opinion.destroy
+
+    redirect_back(fallback_location: articles_path)
+  end
+
+  private
+  def check_for_opinion_collision
```

Comment:
instead of checking for collision it could make sense just automatically destroy previous choice and create new one

Reply:
That was a  very good suggestion. I fixed it!